### DISCUSSION TITLE
How to handle data schemas

### DISCUSSION THREAD

**Discussion by canpacis** [NucleoidAI/Nucleoid#15] [Q&A] [answered]
Can I enforce a property on a class when creating a new record? As far as I can see, there is no way to create a schema for the data. Are you planning to include such behaviour? 

```javascript
class User {
  constructor(name) {
    this.name = name;
  }
}
nucleoid.register(User);

// I need to manually check the fields for every parameter.
app.post("/users", (req) => new User(req.body.name));
```

With this, there is also no way to make a field unique. Am I understanding it correctly?

**Comment by canmingir** [comment, marked as answer]
I am glad that you opened this up in discussions because we are still designing the runtime, in current stage, we are working to integrate with OpenAPI [here](https://github.com/NucleoidJS/Nucleoid/blob/main/src/libs/openapi.js). By the design, when it gets the JavaScript part, data will be already validated according the defined API, and for all of business logic and constraint logic, it is just the pure JavaScript, like for your example, unique constraint, it is going to be look like this:

```javascript
app.post("/users", (req) => {
  const username = req.body.username
    
  if(User[username]) {
    throw "USER_ALREADY_REGISTERED"
  }

  return new User(req.body.username)
});
```

The way that Nucleoid works is like the runtime rerenders JavaScript is passed by, manages control flow as well as `fs`. In short, we can automate lots of part of them, but my concern it will become too opinionated, especially at this stage, we keep the runtime generic as much as possible, but providing tooling libs like lodash maybe help to build simpler constraint logic.
